design-docs/012-python-namespaces

```diff
@@ -0,0 +1,216 @@
+
+## Solution
+
+It would be better practice to split this up into sub-modules. That would improve the user experience in a few ways:
+
+- auto-complete would be more helpful
+- the API docs would be easier to navigate (see also [#1095](https://github.com/earth-mover/icechunk/issues/1095))
+- the key classes would be the only ones in the top-level namespace (e.g. `Repo`, `Session`, `IcechunkStore`), making it more immediately obvious how icechunk is structured
+- the words "store"/"storage" would become a bit less overloaded (e.g. `icechunk.Storage`, `icechunk.s3_store`, and `icechunk.IcechunkStore` are all currently in the same namespace, but mean totally different things)
+
+## Options for namespace groupings
+
+1. **Group by related functionality**
+
+I.e. everything related to storage together, everything related to virtual chunks together...
+
+2. **Group by cloud provider / storage type**
+
+I.e. everything for Azure together, everything for GCS together...
+
+Are there any others?
+
+## User experience for icechunk beginners
+
+We don't want it to be too hard for beginners to find the imports they need to get started. There is therefore a tension between strict modularization and initial ease-of-use.
+
+## Specific proposal with rationale
+
+Here's a specific proposal. The rationale is that opening or creating should be top-level, but otherwise everything else should be in specific modules.
+
+Some opinionated notes about the rationale used:
+- This proposal is aggressive: it reduces the number of things in the top-level namespace from 78 to 5.
+- This proposal chooses to group by related functionality.
+- The location of classes that are only really returned to the user, and rarely directly called by the user (e.g. `Diff`), doesn't really matter.
+  - Similarly classes with no constructor methods (e.g. `Session`) don't need to be in the top-level namespace.
+- Modules with only one member are fine if they add clarity.
+- Adding an extra intermediate namespace doesn't make things much harder for users to find. In other words this:
+
+```python
+import icechunk
+storage = icechunk.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+is not significantly harder than this:
+
+```python
+import icechunk
+storage = icechunk.storage.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+Also note that this is already valid API:
+
+```python
+import icechunk
+storage = icechunk.Storage.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+### Top-level `icechunk` module:
+
+- `icechunk.Repository` (also available in `icechunk.repository.Repository`)
+- `icechunk.IcechunkStore` (also available in `icechunk.store.IcechunkStore`)
```

Comment:
Do people need to use `IcechunkStore` directly? Can be a good candidate to hide in a module

Reply:
yeah, let's move it. Even typechecking would be rare, because you can use Zarr's store